google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py

```diff
@@ -0,0 +1,232 @@
+    """The RuleBook for ExternalProjectAccess resources."""
+
+    # Class variable for matching the ancestor during rule validation
+    ancestor_pattern = re.compile(r'^organizations/\d+$|^folders/\d+$')
+
+    #pylint: disable=W0613
+    def __init__(self, global_configs=None, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+            global_configs (dict): Inventory configurations.
+        """
+        super(ExternalProjectAccessRuleBook, self).__init__()
+        self.resource_rules_map = dict()
+
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book.
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+        try:
+            for (i, rule) in enumerate(rule_defs.get('rules', [])):
+                self.add_rule(rule, i)
+        except TypeError:
+            pass
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        ancestor = ExternalProjectAccessRuleBook.process_rule(rule_def,
+                                                              rule_index)
+        rule = Rule(rule_name=rule_def.get('name'),
+                    rule_index=rule_index,
+                    rules=ancestor)
+        if ancestor not in self.resource_rules_map.keys():
+            self.resource_rules_map[ancestor] = rule
+
+    @classmethod
+    def process_rule(cls, rule_def, rule_index):
+        """Process a rule.
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+
+        Returns:
+            ancestor: The ancestor resource defined in the rule
+        """
+        ancestor = rule_def.get('ancestor', None)
+        cls.validate_ancestor(ancestor, rule_index)
+        return resource_util.create_resource(ancestor.split('/')[1],
+                                             resource_util.type_from_name(ancestor))
+
+    @classmethod
+    def validate_ancestor(cls, ancestor, rule_index):
+        """Validate the ancestor in a rule.
+        Args:
+            ancestor (str): The ancestor defined by the rule.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+
+        """
+
+        if not ancestor or len(ancestor) < 1:
```

Comment:
If ancestor is `str`, how can it be less than 1?  i.e. why should this be checked here?